pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -59,34 +65,62 @@ func NewScheduler(kubeClient client.Client) *Scheduler {
 func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Node, error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
 	constraints := provisioner.Spec.Constraints.DeepCopy()
+	start := time.Now()
+
+	sort.Slice(pods, byCPUAndMemoryDescending(pods))
+	sort.Slice(instanceTypes, func(i, j int) bool {
+		return instanceTypes[i].Price() < instanceTypes[j].Price()
+	})
+
 	// Inject temporarily adds specific NodeSelectors to pods, which are then
 	// used by scheduling logic. This isn't strictly necessary, but is a useful
 	// trick to avoid passing topology decisions through the scheduling code. It
 	// lets us treat TopologySpreadConstraints as just-in-time NodeSelectors.
 	if err := s.topology.Inject(ctx, constraints, pods); err != nil {
 		return nil, fmt.Errorf("injecting topology, %w", err)
 	}
-	return s.schedule(constraints, instanceTypes, pods), nil
-}
 
-// schedule separates pods into a list of TheoreticalNodes that contain the pods. All pods in each theoretical node
-// contain isomorphic scheduling constraints and can be deployed together on the same node, or multiple similar nodes if
-// the pods exceed one node's capacity.
-func (s *Scheduler) schedule(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) []*Node {
-	var nodes []*Node
+	nodes, err := NewNodeSet(ctx, constraints, s.kubeClient)
+	if err != nil {
+		return nil, fmt.Errorf("constructing nodeset, %w", err)
+	}
 
 	for _, pod := range pods {
 		isScheduled := false
-		for _, node := range nodes {
+		for _, node := range nodes.nodes {
 			if err := node.Compatible(pod); err == nil {
 				node.Add(pod)
 				isScheduled = true
 				break
 			}
 		}
 		if !isScheduled {
-			nodes = append(nodes, NewNode(constraints, instanceTypes, pod))
+			nodes.Add(NewNode(constraints, nodes.daemons, instanceTypes, pod))
+		}
+	}
+	logging.FromContext(ctx).Infof("scheduled %d pods onto %d nodes in %s", len(pods), len(nodes.nodes), time.Since(start))
+	return nodes.nodes, nil
+}
+
+func byCPUAndMemoryDescending(pods []*v1.Pod) func(i int, j int) bool {
+	return func(i, j int) bool {
+		lhs := resources.RequestsForPods(pods[i])
+		rhs := resources.RequestsForPods(pods[j])
+
+		cpuCmp := resources.Cmp(lhs[v1.ResourceCPU], rhs[v1.ResourceCPU])
+		if cpuCmp < 0 {
+			// LHS has less CPU, so it should be sorted after
+			return false
+		} else if cpuCmp > 0 {
+			return true
+		}
+		memCmp := resources.Cmp(lhs[v1.ResourceMemory], rhs[v1.ResourceMemory])
+
+		if memCmp < 0 {
+			return false
+		} else if memCmp > 0 {
+			return true
 		}
+		return false
```

Comment:
Consider shortening to inline
```suggestion
		if cpuCmp :=  resources.Cmp(lhs[v1.ResourceCPU], rhs[v1.ResourceCPU]); cpuCmp != 0 {
		    return cpuCmp > 0
		}
		return resources.Cmp(lhs[v1.ResourceMemory], rhs[v1.ResourceMemory]) > 0
```

Reply:
I'm not convinced on this one.  Your version is shorter, but I think it takes less time to read my version, or it takes me less time anyway.

Reply:
I'm with Nick on this one.

Reply:
We'll have to agree to disagree on this one 😆 I can change it if you like, however to me line 116 requires more thought and is easier to misinterpret.